src/cpp/ext/gsm/metadata_exchange.cc

```diff
@@ -0,0 +1,243 @@
+  std::string pod_name_value = GetStringValueFromAttributeMap(
+      map, opentelemetry::sdk::resource::SemanticConventions::kK8sPodName);
+  std::string container_name_value = GetStringValueFromAttributeMap(
+      map,
+      opentelemetry::sdk::resource::SemanticConventions::kK8sContainerName);
+  std::string namespace_value = GetStringValueFromAttributeMap(
+      map,
+      opentelemetry::sdk::resource::SemanticConventions::kK8sNamespaceName);
+  std::string cluster_name_value = GetStringValueFromAttributeMap(
+      map, opentelemetry::sdk::resource::SemanticConventions::kK8sClusterName);
+  std::string cluster_location_value = GetStringValueFromAttributeMap(
+      map, opentelemetry::sdk::resource::SemanticConventions::
+               kCloudRegion);  // if regional
+  if (cluster_location_value == "unknown") {
+    cluster_location_value = GetStringValueFromAttributeMap(
+        map, opentelemetry::sdk::resource::SemanticConventions::
+                 kCloudAvailabilityZone);  // if zonal
+  }
+  std::string project_id_value = GetStringValueFromAttributeMap(
+      map, opentelemetry::sdk::resource::SemanticConventions::kCloudAccountId);
+  std::string canonical_service_value =
+      grpc_core::GetEnv("CANONICAL_SERVICE_NAME").value_or("unknown");
+  // Create metadata to be sent over wire.
+  AddStringKeyValueToStructProto(metadata, kMetadataExchangeTypeKey, type_value,
+                                 arena.ptr());
+  AddStringKeyValueToStructProto(metadata, kMetadataExchangePodNameKey,
+                                 pod_name_value, arena.ptr());
+  AddStringKeyValueToStructProto(metadata, kMetadataExchangeContainerNameKey,
+                                 container_name_value, arena.ptr());
+  AddStringKeyValueToStructProto(metadata, kMetadataExchangeNamespaceNameKey,
+                                 namespace_value, arena.ptr());
+  AddStringKeyValueToStructProto(metadata, kMetadataExchangeClusterNameKey,
+                                 cluster_name_value, arena.ptr());
+  AddStringKeyValueToStructProto(metadata, kMetadataExchangeLocationKey,
+                                 cluster_location_value, arena.ptr());
+  AddStringKeyValueToStructProto(metadata, kMetadataExchangeProjectIdKey,
+                                 project_id_value, arena.ptr());
+  AddStringKeyValueToStructProto(metadata,
+                                 kMetadataExchangeCanonicalServiceNameKey,
+                                 canonical_service_value, arena.ptr());
+  size_t output_length;
+  char* output =
+      google_protobuf_Struct_serialize(metadata, arena.ptr(), &output_length);
+  serialized_labels_to_send_ = grpc_core::Slice::FromCopiedString(
+      absl::Base64Escape(absl::string_view(output, output_length)));
+  // Fill up local labels map. The rest we get from the detected Resource and
+  // from the peer.
+  local_labels_.emplace_back("canonical_service_name", canonical_service_value);
```

Comment:
I think we don't actually need to add canonical_service_name as a local label, because that will be part of the MonitoredResource, right?  I think we just need to record the remote canonical_service_name.